### ISSUE TITLE
Request ID middleware

### ISSUE THREAD

**Issue by jnichols-git** [cloudretic/matcha#106]
Should've done this a long time ago. Ideally, should define a function type that generates IDs and a middleware that takes that type and assigns the result to each inbound request, either in context or as a header (configurable by user).

**Comment by mdrokz** [comment]
> Should've done this a long time ago. Ideally, should define a function type that generates IDs and a middleware that takes that type and assigns the result to each inbound request, either in context or as a header (configurable by user).

Hi can i work on this issue ? and can you guide me in implementing this ?

**Comment by jnichols-git** [comment]
@mdrokz Absolutely! Assigning now.

Quick implementation notes:

- This should be implemented in `pkg/middleware` in a unique file, name up to you.
- Most middleware implementations follow this pattern:

```
func MiddlewareName([params here]) Middleware {
     return func(w http.ResponseWriter, req *http.Request) *http.Request {
        [implementation here]
     }
}
```

- You can either make the parameters control if the ID is assigned to context or a header, or make two separate functions. Two separate functions is probably easier.
- The parameter name in context, or header name, should also be user choice.

I've got to get to work but if you have more questions drop by our Discord here: <https://discord.gg/fKWmGamz> I've just been made aware the invite on the repo is invalid, so I'll get that fixed.

**Comment by mdrokz** [comment]
> @mdrokz Absolutely! Assigning now.
> 
> Quick implementation notes:
> 
> - This should be implemented in `pkg/middleware` in a unique file, name up to you.
> - Most middleware implementations follow this pattern:
> 
> ```
> func MiddlewareName([params here]) Middleware {
>      return func(w http.ResponseWriter, req *http.Request) *http.Request {
>         [implementation here]
>      }
> }
> ```
> 
> - You can either make the parameters control if the ID is assigned to context or a header, or make two separate functions. Two separate functions is probably easier.
> - The parameter name in context, or header name, should also be user choice.
> 
> I've got to get to work but if you have more questions drop by our Discord here: <https://discord.gg/fKWmGamz> I've just been made aware the invite on the repo is invalid, so I'll get that fixed.
> 

Thanks for all of the details I will start on it in the morning and I will check out the discord server. Thanks!

**Comment by jnichols-git** [comment]
Closed as complete by #111
